pkg/config/config.go

```diff
@@ -95,63 +104,116 @@ func LoadConfig(lokocfgPath, lokocfgVarsPath string) (*HCLConfig, hcl.Diagnostic
 		}
 	}
 
-	hclParser := hclparse.NewParser()
-
-	var hclFiles []*hcl.File
-	for _, f := range lokocfgPaths {
-		hclFile, diags := hclParser.ParseHCLFile(f)
-		if len(diags) > 0 {
-			return nil, diags
-		}
-		hclFiles = append(hclFiles, hclFile)
-	}
-
-	configBody := hcl.MergeFiles(hclFiles)
-
-	exists, err := util.PathExists(lokocfgVarsPath)
-	if err != nil {
+	if len(lokocfgPaths) == 0 {
 		return nil, hcl.Diagnostics{
 			&hcl.Diagnostic{
 				Severity: hcl.DiagError,
-				Summary:  fmt.Sprintf("could not stat %q: %v", lokocfgVarsPath, err),
+				Summary:  "could not find any `.lokocfg` files in the provided path",
 			},
 		}
 	}
-	var userVals map[string]cty.Value
-	var diags hcl.Diagnostics
-	if exists {
-		userVals, diags = LoadValuesFile(lokocfgVarsPath)
-		if len(diags) > 0 {
-			return nil, diags
+
+	for _, path := range lokocfgPaths {
+		data, err := loadHCLFile(path)
+		if err != nil {
+			return nil, hcl.Diagnostics{
+				&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  err.Error(),
+				},
+			}
 		}
+
+		files[path] = data
 	}
 
-	var clusterConfig ClusterConfig
-	diags = gohcl.DecodeBody(configBody, nil, &clusterConfig)
-	if len(diags) > 0 {
-		return nil, diags
+	return files, nil
+}
+
+func loadHCLFile(path string) ([]byte, error) {
+	data, err := ioutil.ReadFile(filepath.Clean(path))
+	if err != nil {
+		return nil, err
 	}
 
+	return data, nil
+}
+
+func getHCLFiles(files map[string][]byte) ([]*hcl.File, hcl.Diagnostics) {
+	var diagnostics hcl.Diagnostics
+
+	hclFiles := []*hcl.File{}
+
+	hclParser := hclparse.NewParser()
+
+	for path, content := range files {
+		f, diags := hclParser.ParseHCL(content, path)
+		diagnostics = append(diagnostics, diags...)
+		hclFiles = append(hclFiles, f)
+	}
+
+	return hclFiles, diagnostics
+}
+
+func parseVariables(userVals map[string]cty.Value, cfg ClusterConfig) (map[string]cty.Value, hcl.Diagnostics) {
+	var diagnostics hcl.Diagnostics
+
 	variables := map[string]cty.Value{}
-	for _, v := range clusterConfig.Variables {
+
+	for _, v := range cfg.Variables {
 		if userVal, ok := userVals[v.Name]; ok {
 			variables[v.Name] = userVal
 			continue
 		}
+
 		if len(v.Default) == 0 {
 			continue
 		}
+
 		defaultValue, hasDefaultValue := v.Default["default"]
 		if !hasDefaultValue {
 			continue
 		}
 		defaultVal, diags := defaultValue.Expr.Value(nil)
-		if len(diags) > 0 {
-			return nil, diags
-		}
+		diagnostics = append(diagnostics, diags...)
+
 		variables[v.Name] = defaultVal
 	}
 
+	return variables, diagnostics
+}
+
+// ParseHCLFiles parses the HCL files into an instance of Config.
+func ParseHCLFiles(configFiles, variableFiles map[string][]byte) (*HCLConfig, hcl.Diagnostics) {
+	var diagnostics hcl.Diagnostics
+
+	configHCLFiles, diags := getHCLFiles(configFiles)
+	diagnostics = append(diagnostics, diags...)
```

Comment:
I wonder if we should do the error checking somewhere earlier. Does all further function calls work if e.g. you don't have permission to read the file?

Reply:
If there is no permission to read the file, an appropriate message is shown to the user.

```
~/go/src/github.com/kinvolk/lokomotive/lokoctl cluster apply
FATA[0000] <nil>: open cluster.lokocfg: permission denied;   args="[]" command="lokoctl cluster apply"
```